lib/linked_list.rb

```diff
@@ -18,71 +21,213 @@ def initialize
+      current = @head 
+      if @head.nil?
+        return nil
+      elsif @head.next.nil?
+        return @head.data
+      else 
+        while current.next != nil
+          if current.data < min
+            min = current.data
+          end
+          current = current.next
+        end 
+      end 
+
+      if current.data < min
+        min = current.data
+      end
+
+      return min
     end
 
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(N)
+    # Space Complexity: O(1)
     def length
-      raise NotImplementedError
+      length = 1
+      current = @head
+      if current.nil?
+        return 0
+      elsif current.next.nil?
+        return 1
+      else 
+        while current.next != nil
+          length +=1
+          current = current.next
+        end
+      end
+      return length
     end
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns nil if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(N)
+    # Space Complexity: O(1)
     def get_at_index(index)
-      raise NotImplementedError
+      count = 0 
+      if index > self.length-1
+        return nil
+      end 
+
+      current = @head
+      if index == 0 
+        return @head.data
+      else
+        while current.next != nil
+          current = current.next
+          count += 1
+          if count == index
+            return current.data
+          end
+        end
+      end
+      if index == self.length-1
+        return current.data
+      end
     end
 
     # method to print all the values in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(N)
+    # Space Complexity: O(1)
     def visit
-      raise NotImplementedError
+      current = @head
+      while current.next != nil
+        puts current.value
+        current = current.next
+      end 
+      puts current.value
     end
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(N)
+    # Space Complexity: O(1)
     def delete(value)
-      raise NotImplementedError
+      if @head.nil? 
+        return nil
+      end 
+      current = @head 
+      while current != nil
+        if @head.next.nil? && @head.data == value
+          @head = nil 
+        elsif @head.next != nil && @head.data == value
+          @head = @head.next
+        elsif current.next != nil && current.prev != nil && current.data == value
+            prevn = current.prev
+            nextn = current.next
+            prevn.next = nextn
+            nextn.prev = prevn
+        elsif  !current.prev.nil? && current.data == value
+          current.prev.next = current.next 
+        end 
+        current = current.next
+      end

```

Comment:
This is a bit overcomplicated

```suggestion
      if @head.data == value
        @head = @head.next
        return
      end
      current = @head
      previous = nil
      
      until current.nil?
        if current.data == value
          previous.next = current.next
          return
        end
        
        previous = current
        current = current.next
      end
      

```